flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/WorkerResourceSpec.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.resourcemanager;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.resources.CPUResource;
+import org.apache.flink.configuration.MemorySize;
+import org.apache.flink.runtime.clusterframework.TaskExecutorProcessSpec;
+import org.apache.flink.util.Preconditions;
+
+import java.util.Objects;
+
+/**
+ * Resource specification of a worker, mainly used by SlotManager requesting from ResourceManager.
```

Comment:
suggestion: Document `null` semantic of `WorkerResourceSpec`.

Reply:
Good point.
I think it might be better to document it in `SlotManagerImpl` rather than `WorkerResourceSpec`. WDYT?

Reply:
My concern is that we have many `@Nullable` `WorkerResourceSpec` parameters. The semantic of `null` `WorkerResourceSpec` belongs to itself. So we can document the design in the class doc of `WorkerResourceSpec` and linked to this universe doc where a nullable param in use.

Reply:
I tend to think it differently.
I think the semantic of a `WorkerResourceSpec` being `null` belongs to the context where it is used, not the class itself. E.g., when we create the `SlotManagerImpl`, the `WorkerResourceSpec` means it is on a standalone setup and SM should not request any worker from RM. On the other hand, when we create a `TaskExecutorProcessSpec` from the `WorkerResourceSpce`, it must not be `null`.

Reply:
Yes that makes sense. Nullable return value and field has their meaning respectively. It seems we have too many Nullable WorkerResourceSpec parameter(just for passing from here to there), maybe they will be later gathered into one single source?

Reply:
True, we have the `@Nullable WorkerResourceSpec` along the code path for creating the `SlotManagerImpl`. To be specific, in the following classes.
- `ResourceManagerRuntimeServiceConfiguration`
- `SlotManagerConfiguration`
- `SlotManagerImpl`
- `SlotManagerBuilder` (test-only)

I guess the root cause is that we decide `WorkerResourceSpec` differently (whether `null` and the `cpu` config options) across different RMs, so we have to pass it from the `ResourceManagerFactory`. Unfortunately, ATM I don't see a good way to resolve that. Do you have any suggestion on that?

Reply:
There are too many `XxxConfiguration` which make the call chain much longer. We might 
 refactor to reduce some poor `XxxConfiguration` later. But it doesn't have to be included in this PR.

Reply:
Or in this pr you can try to directly pass `WorkerResourceSpec` to service's  & manager's `fromConfiguration` instead of configuration's & build's.

Reply:
I'm not sure why the `XxxConfiguration` is used at the first place. Probably there are some good reasons. From what I see, the `XxxConfiguration` clearly shows which information is needed for creating the `Xxx`.
I think we can discuss the pros and cons of `XxxConfiguration` in another thread, maybe a code style discussion in ML? For this PR, I would also like to exclude it from the scope.